src/LSTM/index.js

```diff
@@ -87,55 +85,69 @@ class LSTM {
     };
 
     for (let i = 0; i < this.cellsAmount; i += 1) {
-      c.push(tf.zeros([1, this.model[`Bias_${i}`].shape[0] / 4]));
-      h.push(tf.zeros([1, this.model[`Bias_${i}`].shape[0] / 4]));
-      LSTMCells.push(lstm(i));
+      this.zeroState.c.push(tf.zeros([1, this.model[`Bias_${i}`].shape[0] / 4]));
+      this.zeroState.h.push(tf.zeros([1, this.model[`Bias_${i}`].shape[0] / 4]));
+      this.cells.push(lstm(i));
     }
 
-    const userInput = Array.from(seed);
+    this.state = this.zeroState;
+  }
 
+
+  async generateInternal(options) {
+    await this.ready;
+    const seed = options.seed || this.defaults.seed;
+    const length = +options.length || this.defaults.length;
+    const temperature = +options.temperature || this.defaults.temperature;
+    const stateful = options.stateful || this.defaults.stateful;
+    if (!stateful) {
+      // await this.initCells();
+      this.state = this.zeroState;
+    }
+
+    const results = [];
+    const userInput = Array.from(seed);
     const encodedInput = [];
+
     userInput.forEach((char, ind) => {
-      if (ind < 100) {
+      if (ind < 100) { // TODO: What is this?
         encodedInput.push(this.vocab[char]);
       }
     });
 
-    let current = 0;
-    let input = encodedInput[current];
+    let input = encodedInput[0];
+    let probabilitiesNormalized = []; // will contain final probabilities (normalized)
 
-    for (let i = 0; i < userInput.length + length; i += 1) {
+    for (let i = 0; i < userInput.length + length + -1; i += 1) {
```

Comment:
this is easier to explain with an example (also for my own sanity check :)

let's say the seed is `WOR`and we want to generate 1 character
```
userInput.length == 3
length == 1
userInput.length + length -1 -> 3
```
We loop:
0: feed ('W', zeroState), fetch (state, probabilities)
1: feed ('O', state), fetch (state, probabilities) 
2: feed ('R', state), fetch (state, probabilities), sample -> e.g. 'D' 

Now this is perhaps up to debate: I think the lstm function should return the sample 'D', and the probabilities for the current last step (i.e. that 'D' was sampled from), so that an app has the option to ignore the 'D' and take a different sample if need be (i.e. the lstm function returns the  probability distribution as a result of feeding in `WOR`, which is the seed). If we were to loop once more and feed the 'D' back in, we would get the probabilities for the next step (i.e. having fed in `WORD`), I don't think this would be the desired behaviour. 

A slightly debatable aspect though, is of the state. With the loop as it currently is, the state that is returned, is before the last sample is fed in. From a certain perspective, this makes sense: You seed with `WOR`, and you get the results as a probability distribution having fed in `WOR`, and the state having fed in `WOR`, *not* having fed in `WORD`.

For most use cases (i.e. stateless LSTM, or if not using the probability distribution), none of this makes any difference. It's only when you want to use stateful lstm, or access the probabilities manually, does this become important. And I realise this may be a bit confusing for people. Basically right now using state requires a bit of knowledge and understanding what's going on. I don't know if there's a way we can make this easier and less confusing?

I thought about adding an `autoSample` option (a bit similar to what you suggest below), so that

```
if autoSample == true:
    sample final probability distribution (i.e. having fed `WOR`) -> e.g. 'D'
    feed 'D' back in to update state (to that of being fed 'WORD')
    do *not* sample the new probability distribution (we don't need another character)
    return sample (e.g. 'D') (internal state is that of being fed 'WORD')
else :
   current behaviour (but don't even bother sampling the final step - we don't need it),
   return probabilities (having fed in WOR, internal state is also that of being fed WOR)
   (We actually don't ever need to return the state, since it's a class property).
```
But I decided against this because I thought it would be more confusing if the final internal state is different whether you use autoSample or not!!! Also, if somebody wanted to use autosample for ease of use, but also visualise the probabilities, they'd need to use autoSample==True, but the probabilities would be showing the wrong (i.e.) next step.

P.S. If we wanted to generate two characters, same logic applies I think:
We loop:
0: feed ('W', zeroState), fetch (state, probabilities)
1: feed ('O', state), fetch (state, probabilities) 
2: feed ('R', state), fetch (state, probabilities), sample -> e.g. 'D' 
3: feed ('D', state), fetch (state, probabilities), sample -> e.g. 'S' 
return probabilities from which 'S' was sampled, and the state having seeded 'WORD'.


(P.P.S. I had to add -1, because I couldn't figure out how to subtract 1, with the crazy super strict checks of the builder! :)

Reply:
thanks for the detailed explanation!
A couple of notes:
- Let's return the probabilities until the current step. (until `WOR` not having fed `D`) and mention this with detail in the docs. Not sure about enabling `autoSample`, as you said it might be more confusing. 
- What if we parse the probabilities in an easier/friendly way? Like returning an object with  the 10 most likely chars instead of the complete vocab with all probs?
- With the addition of the `setState` and `getState`, not returning the state in `generated` makes sense.